### DISCUSSION TITLE
Usage of x-for with an undeclared object

### DISCUSSION THREAD

**Discussion by oscarmcm** [alpinejs/alpine#3323] [7. Bugs (archived)] [answered]
Hi, I'm trying to port a project from AlpineJs 2.X to the 3.X (using 3.10.5) and I have a template that is closer to this:

```html
<ul x-data="myComponent()">
  <template x-for="(item, index) in items" :key="index">
    <li x-text="item.name"></li>
  </template>
</ul>

```

```js
function myComponent() {
  return {
    items: [],
    init() {
     var _this = this;
     $.get('some_dataset', function(response) {
       _this.items = response['items'];
     });
    }
  }
};
```

This did not happen on Alpine v2 but on the new version, it does render a console error:

```txt
[Warning] Alpine Expression Error: Can't find variable: item (alpine.min.js, line 1)

Expression: "item.name"


<li x-text="item.name"></li>
```

And the only way to get rid of the error is declaring a new object with the schema to be used in the template:

```diff
function myComponent() {
  return {
    items: [],
+   item: {name: ''},  
    init() {
     var _this = this;
     $.get('some_dataset', function(response) {
       _this.items = response['items'];
     });
    }
  }
};
```

Is there a new way to achieve the same without declaring a base object in the template? 

Thanks in advance.

**Comment by SimoTod** [comment]
The code you posted works okay for me: https://codepen.io/SimoTod/pen/zYaeXqy?editors=1111
Can you replicate the error somewhere and post the link?

**Reply by oscarmcm** [reply]
hey @SimoTod thanks for the quick answer, I was able to replicate it here: https://jsfiddle.net/s0c69vbo/1/

Click on the get data button and it will show the warning in the console, and after testing this seems to happen if the `<template>` has an `x-if`

**Reply by SimoTod** [reply]
I believe there are issues in v3 when you use x-if and x-for on the same template tag. 
Regarding your code, I don't think you need x-if at all, x-for won't have anything to loop on anyway so the if condition on the length is implicit. If you really want it, it has to be something like ` <template x-if="users.length > 0"><template x-for="(user, index) in users" :key="index">`

**Comment by ekwoka** [comment, marked as answer]
You can't use x-if and x-for together.

What I recommend in those use cases is just do the x-for with an expression of `(item, index) in condition ? items : []`.

In your example that is showing the issue, the condition isn't needed at all, as a loop of an empty array does nothing. So if you're only checking if the array is not empty, that's already built into how array loops work.

also, I highly recommend throwing out that nasty jquery.

**Comment by oscarmcm** [comment]
Thanks @ekwoka and @SimoTod for the answers.
